### ISSUE TITLE
Update ArgumentsAreDifferent for JUnit4+

### ISSUE THREAD

**Issue by kriegfrj** [mockito/mockito#1649]
From #1642:
>There is code in `org.mockito.internal.junit.ExceptionFactory` that checks to see if JUnit is on the classpath - if so, then it will throw `org.mockito.exceptions.verification.junit.ArgumentsAreDifferent` rather than `org.mockito.exceptions.verification.ArgumentsAreDifferent`. The JUnit version inherits from `junit.framework.ComparisonFailure` whereas the basic version inherits from `org.mockito.exceptions.base.MockitoAssertionError`.
>
>...it would make more sense to derive from a more-up-to-date version of `ComparisonFailure` if it is available, like `org.junit.ComparisonFailure` - perhaps this could be the subject of a different issue/PR)

This is that issue. Create another ArgumentsAreDifferent exception that derives from `org.junit.ComparisonFailure` rather than `junit.framework.ComparisonFailure`, and then use the most up-to-date-one that is available on the classpath. The up-to-date one has better reporting and diagnostics.

One minor irritant with this might be that if you are using Mockito with JUnit 4+ on the classpath but with the JUnit 3 runner, your Mockito assertions will be reported as errors instead of failures. However, I think that would be a pretty odd combination of tests/runners to use and probably not worth worrying about. The alternative would be to dynamically check which runner is being used, which is doable but a bit more tricky.

**Comment by mockitoguy** [comment]
@kriegfrj, why do we need this change? JUnit 4.x development is complete, and there won't be any future versions.

**Comment by mockitoguy** [comment]
I suggest we take the discussion to the ticket: https://github.com/mockito/mockito/issues/1649#issuecomment-469742700

Thank you for attempting to tackle this!

**Comment by mockitoguy** [comment]
Unless we have evidence, I don't see any benefit of changing anything for JUnit 4.x, which is no longer developed. However, it would be cool to have this working for JUnit5 :)

**Comment by kriegfrj** [comment]
>@kriegfrj, why do we need this change? JUnit 4.x development is complete, and there won't be any future versions.

Well, "need" is a relative term. 😄

`org.junit.ComparisonFailure` is obviously newer than `junit.framework.ComparisonFailure`. It had some extra features (I think mostly related to readability of the output) that weren't back-ported into `junit.framework.ComparisonFailure`. So the need isn't great; however I (naively, it seems) thought it would be a straightforward, incremental improvement to switch from one to the other. It was not until I had to refactor a bunch of Mockito's own test cases (the ones that caught `.junit.ArgumentsAreDifferent`) that I realised it was a bit bigger than I thought.

>Unless we have evidence, I don't see any benefit of changing anything for JUnit 4.x, which is no longer developed. However, it would be cool to have this working for JUnit5 :)

Well, that opens up a whole other can of worms. 😄 

There are still quite a few tests out there that are running using the JUnit 4 runner, even though it is no longer developed. I'm one of those who has been too lazy to switch to JUnit 5 for most of my own projects, and all the 3rd-party projects I've worked on lately (bnd, assertj, mockito) are still using the JUnit 4. I had also (again naively, it seems) assumed that JUnit 5 was still using `org.junit.ComparisonFailure` - hence the name of this issue was "...for JUnit4**+**" as I originally thought this would fix it for both. But it seems that JUnit 5 instead initiated the Open Test Alliance for the JVM project (https://github.com/ota4j-team/opentest4j) and is using `org.opentest4j.AssertionFailedError`.

So perhaps the full implementation of this ticket would be to include support for both. This would be trivial to add (in fact, I've already done it in my working copy); however whether we add support for either or both, the hardest part will be how to do this while minimising breaking changes. I think I'll start a new post for that discussion.

**Comment by kriegfrj** [comment]
### Managing backward compatibility
As it turns out, the biggest issue with this... errrmm... issue, is how best to maintain backward compatibility. I outlined the options here: https://github.com/mockito/mockito/pull/1653#issue-258371083.

@TimvdLippe pointed out (https://github.com/mockito/mockito/pull/1653#issuecomment-469727148), changing the package name of the `ArgumentsAreDifferent` exception is a breaking change.

However, changing the inheritance hierarchy of `org.mockito.exceptions.verification.junit.ArgumentsAreDifferent` would also be a breaking change (as I discovered) - it's a matter of which change will break less.

The problem (as I highlighted in the PR) is that there may be two classes of code out there:

1. Code that relies on the fact that, when running under JUnit, Mockito generates a `junit.framework.ComparisonFailure` exception.
2. Code that relies on the fact that, when running under JUnit, Mockito generates a
`org.mockito.exceptions.verification.junit.ArgumentsAreDifferent` exception.

Having pondered this for some time now, I think the only way to maintain 100% backward compatibility when adding this feature (particularly for code in the first class) is to have a global configuration flag to enable it. Making it disabled by default will preserve backward compatibility. However, this would present a "barrier to entry" and might mean that code doesn't migrate to use the new exception.

So I think the two best options are:

1. Keep `org.mockito.exceptions.verification.junit.ArgumentsAreDifferent` as the "most current" version. This will cause the least breakage of existing code - it breaks binary compatibility but keeps source compatibility (so a recompile will be necessary to use the latest version). Only code that attempts to catch `junit.framework.ComparisonFailure` will be broken & need updating. Gets complicated if we try and support all three levels of exception.
2. Use a flag to turn the feature on. off by default. 100% backward compatibility (binary and source), but presents a barrier to use.
3. Use a flag to turn feature on, on by default. Eliminates the barrier to use, easy migration path for anyone with legacy code.

I think the decision on which way to go is probably above my "pay grade" for this project - I'll leave it to the core developers. I'm tending towards the third solution as I'm guessing the amount of code out there using a JUnit 4 or 5 runner and attempting to catch the legacy `ComparisonFailure` is small, and the migration path is easy enough for those cases.

Thoughts welcome.

**Comment by kriegfrj** [comment]
Btw: any comment on my question about source level compatibility? As I noted in my PR and in comments, the source of `ExceptionFactory` could be made substantially cleaner through the use of Java 8 method handles, but unless I'm mistaken we're still on an earlier source level?

**Comment by mockitoguy** [comment]
Hey @kriegfrj!

We don't want to add new configuration because it adds complexity. There is little/no benefit of refactoring JUnit4 exceptions. Let's focus on JUnit5. What we want is good experience in IDEA when JUnit5 is in use, without regressing the experience for JUnit4/other framework users. Can we do a reflection trick and instantiate JUnit5/OpenTest4J comparison failure?

We are not yet on Java8 source compatibility.

Thank you for working on this!

**Comment by kriegfrj** [comment]
> Hey @kriegfrj!

Hey @mockitoguy! 

> We don't want to add new configuration because it adds complexity.

Ok, scratch that option.

>There is little/no benefit of refactoring JUnit4 exceptions. Let's focus on JUnit5. What we want is good experience in IDEA when JUnit5 is in use, without regressing the experience for JUnit4/other framework users. Can we do a reflection trick and instantiate JUnit5/OpenTest4J comparison failure?

It is trivial to do the same reflection trick and instantiate OpenTest4J comparison failure. I have already implemented it in my working copy. The tricky part is that this will break backward compatibility.

I have put this

> We are not yet on Java8 source compatibility.

Is there a reason for holding off? Java 8 really tidies up the code for this significantly (and makes it a bit more efficient too).

> Thank you for working on this!

You're welcome!

**Comment by mockitoguy** [comment]
Let's keep Java8 separate from this PR.

> The tricky part is that this will break backward compatibility.

Can you describe the incompatibility? Is it that we will be throwing a different kind of AssertionError?

**Comment by kriegfrj** [comment]
> Let's keep Java8 separate from this PR.

Done.

> Can you describe the incompatibility? Is it that we will be throwing a different kind of AssertionError?

Yes, because it will change the type of ArgumentsAreDifferent

**Comment by mockitoguy** [comment]
Thank you for clarifying.
